### PR TITLE
[Tune] Respect default_resources during Trial.reset().

### LINKED ISSUE #16891: [tune] Resuming errored trials does not work
### What is the problem?
`tune.run(..., resume='ERRORED_ONLY')` is supposed to reset and rerun failed experiments. It does not do so.

Verified using ray nightly wheel installed 7 July
### Reproduction (REQUIRED)
```
import gym
import ray
from ray import tune
import numpy as np


class TestEnv(gym.Env):
    def __init__(self, cfg={}):
        self.observation_space = gym.spaces.Discrete(1)
        self.action_space = gym.spaces.Discrete(1)

    def reset(self):
        return 0

    def step(self, act):
        raise Exception()

try:
    tune.run("PPO", config={"env": TestEnv, "framework": "torch"})
except:
    pass
tune.run("PPO", config={"env": TestEnv, "framework": "torch"}, resume="ERRORED_ONLY")
```

Truncated output:
```
...
2021-07-06 17:16:09,555 INFO trial_runner.py:308 -- Restarting experiment.
== Status ==
Memory usage on this node: 20.5/187.6 GiB                                                                                                                                                                        Using FIFO scheduling algorithm.
Resources requested: 0/96 CPUs, 0/4 GPUs, 0.0/132.82 GiB heap, 0.0/31.83 GiB objects (0.0/1.0 accelerator_type:RTX)



2021-07-06 17:16:09,660 INFO tune.py:545 -- Total run time: 0.12 seconds (0.00 seconds for the tuning loop).
2021-07-06 17:16:09,661 WARNING experiment_analysis.py:719 -- No `self.trials`. Drawing logdirs from checkpoint file. This may result in some information that is out of sync, as checkpointing is periodic.
Traceback (most recent call last):
  File "test.py", line 14, in <module>
    tune.run("PPO", config={"env": TestEnv, "framework": "torch"}, resume="ERRORED_ONLY")
  File "/usr/local/lib/python3.8/dist-packages/ray/tune/tune.py", line 555, in run
    return ExperimentAnalysis(
  File "/usr/local/lib/python3.8/dist-packages/ray/tune/analysis/experiment_analysis.py", line 425, in __init__
    super(ExperimentAnalysis, self).__init__(
  File "/usr/local/lib/python3.8/dist-packages/ray/tune/analysis/experiment_analysis.py", line 77, in __init__
    self.fetch_trial_dataframes()
  File "/usr/local/lib/python3.8/dist-packages/ray/tune/analysis/experiment_analysis.py", line 201, in fetch_trial_dataframes
    for path in self._get_trial_paths():
  File "/usr/local/lib/python3.8/dist-packages/ray/tune/analysis/experiment_analysis.py", line 726, in _get_trial_paths
    raise TuneError("No trials found.")
ray.tune.error.TuneError: No trials found.
```

If the code snippet cannot be run by itself, the issue will be closed with "needs-repro-script".

- [x] I have verified my script runs in a clean environment and reproduces the issue.
- [x] I have verified the issue also occurs with the [latest wheels](https://docs.ray.io/en/master/installation.html).

### PR DESCRIPTION
**Pull request by xwjiang2010**
## Why are these changes needed?

During trial reset, default_resources take precedence over resources/placement_group_factory. Clear the latter two if default_resources is present.

## Related issue number
Closes #16891

## Checks

- [X] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [X] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/trial.py [hunk 1]
```diff
@@ -404,15 +404,20 @@ def uses_placement_groups(self):
         return bool(self.placement_group_factory)
 
     def reset(self):
+        trainable_cls = self.get_trainable_cls()
```
Comment (matthewdeng) [inline python/ray/tune/trial.py:407]: Nit: I would move the reset logic closer, remove the negation, and add a comment to explain this interaction.

```python
# Unset resources and placement_group_factory if there is a default_resource_request.
# The new Trial will be configured using the default_resource_request.
trainable_cls = self.get_trainable_cls()
clear_resources = (trainable_cls and
                   trainable_cls.default_resource_request(self.config))
resources = None if clear_resources else self.resources 
placement_group_factory = None if clear_resources else self.placement_group_factory

...

resources=resources
placement_group_factory=placement_group_factory
```

**Review by matthewdeng — APPROVED** [review]
LGTM!

For additional context here's the stacktrace of the original problem:

```
2021-09-01 17:09:22,887	WARNING trial_runner.py:510 -- Attempting to resume experiment from /Users/matt/ray_results/PPO. This will ignore any new changes to the specification.
2021-09-01 17:09:22,893	ERROR trial_runner.py:306 -- Resources for <class 'ray.rllib.agents.trainer_template.PPO'> have been automatically set to <ray.tune.utils.placement_groups.PlacementGroupFactory object at 0x7ffa73147ca0> by its `default_resource_request()` method. Please clear the `resources_per_trial` option.
2021-09-01 17:09:22,893	ERROR trial_runner.py:307 -- Runner restore failed.
Traceback (most recent call last):
  File "/Users/matt/workspace/ray/python/ray/tune/trial_runner.py", line 302, in __init__
    self.resume(run_errored_only=errored_only)
  File "/Users/matt/workspace/ray/python/ray/tune/trial_runner.py", line 534, in resume
    new_trial = trial.reset()
  File "/Users/matt/workspace/ray/python/ray/tune/trial.py", line 408, in reset
    return Trial(
  File "/Users/matt/workspace/ray/python/ray/tune/trial.py", line 246, in __init__
    raise ValueError(
ValueError: Resources for <class 'ray.rllib.agents.trainer_template.PPO'> have been automatically set to <ray.tune.utils.placement_groups.PlacementGroupFactory object at 0x7ffa73147ca0> by its `default_resource_request()` method. Please clear the `resources_per_trial` option.
2021-09-01 17:09:22,895	INFO trial_runner.py:310 -- Restarting experiment.
```

File: python/ray/tune/tests/test_api.py [hunk 2]
```diff
@@ -695,6 +696,45 @@ def test(config):
...
+            def reset(self):
+                return 0
+
+            def step(self, act):
+                if self.should_fail:
+                    raise Exception()
+                else:
+                    return [0, 1, True, {}]
+
+        try:
```
Comment (krfricke) [inline python/ray/tune/tests/test_api.py:716]:
```suggestion
        with self.assertRaises(Exception):
```

**Review by krfricke — APPROVED** [review]
Only one nit, otherwise LGTM

**Comment by xwjiang2010** [comment]
Thanks for the review!

**Pull request merged**